native/cocos/renderer/pipeline/shadow/CSMLayers.cpp

```diff
@@ -0,0 +1,300 @@
+    const Vec3 &center = _castLightViewBoundingBox.getCenter();
+    Vec3 shadowPos(center.x, center.y, center.z + r + invisibleOcclusionRange);
+    shadowPos.transformMat4(shadowPos, matShadowTrans);
+
+    Mat4::fromRT(dirLight->getNode()->getRotation(), shadowPos, &matShadowTrans);
+    matShadowView = matShadowTrans.getInversed();
+
+    if (!isOnlyCulling) {
+        // snap to whole texels
+        const float halfOrthoSizeWidth = orthoSizeWidth * 0.5F;
+        const float halfOrthoSizeHeight = orthoSizeHeight * 0.5F;
+        Mat4 matShadowProj;
+        Mat4::createOrthographicOffCenter(-halfOrthoSizeWidth, halfOrthoSizeWidth, -halfOrthoSizeHeight, halfOrthoSizeHeight,
+                                          0.1F, _shadowCameraFar, clipSpaceMinZ, projectionSinY, 0U, &matShadowProj);
+
+        Mat4 matShadowViewProjArbitaryPos;
+        Mat4::multiply(matShadowProj, shadowViewArbitaryPos, &matShadowViewProjArbitaryPos);
+        Vec3 projPos;
+        projPos.transformMat4(shadowPos, matShadowViewProjArbitaryPos);
+        const float invActualSize = 2.0F / shadowMapWidth;
+        const Vec2 texelSize(invActualSize, invActualSize);
+        const float modX = fmodf(projPos.x, texelSize.x);
+        const float modY = fmodf(projPos.y, texelSize.y);
+        const Vec3 projSnap(projPos.x - modX, projPos.y - modY, projPos.z);
+        const Mat4 matShadowViewProjArbitaryPosInv = matShadowViewProjArbitaryPos.getInversed();
+        Vec3 snap;
+        snap.transformMat4(projSnap, matShadowViewProjArbitaryPosInv);
+
+        Mat4::fromRT(dirLight->getNode()->getRotation(), snap, &matShadowTrans);
+        matShadowView = matShadowTrans.getInversed();
+
+        // fill data
+        Mat4 matShadowViewProj;
+        Mat4::multiply(matShadowProj, matShadowView, &matShadowViewProj);
+        _matShadowView = matShadowView.clone();
+        _matShadowProj = matShadowProj.clone();
+        _matShadowViewProj = matShadowViewProj.clone();
+    }
+    _validFrustum.createOrtho(orthoSizeWidth, orthoSizeHeight, 0.1F, _shadowCameraFar, matShadowTrans);
+}
+
+CSMLayerInfo::CSMLayerInfo(uint level) : ShadowTransformInfo(level) {
+    _splitCameraNear = 0.0F;
+    _splitCameraFar = 0.0F;
+    _matShadowAtlas.setZero();
+    _matShadowViewProjAtlas.setZero();
+    calculateAtlas(level);
+}
+
+void CSMLayerInfo::calculateAtlas(uint level) {
+    const gfx::Device *device = gfx::Device::getInstance();
+    const float clipSpaceSignY = device->getCapabilities().clipSpaceSignY;
+    const float x = floorf(static_cast<float>(level % 2U)) - 0.5F;
+    const float y = clipSpaceSignY * (0.5F - floorf(static_cast<float>(level) / 2U));
+    const Vec3 bias(x, y, 0.0F);
+    const Vec3 scale(0.5F, 0.5F, 1.0F);
+    Mat4::fromRTS(Quaternion::identity(), bias, scale, &_matShadowAtlas);
+}
+
+CSMLayers::CSMLayers() {
+    _levelCount = 0U;
+    _specialLayer = new ShadowTransformInfo(1U);
+    _shadowDistance = 0.0F;
```

Comment:
Use initialize _levelCount and _shadowDistance when declaration? The don't have to initialize them here.